Recorder: dump resources to a subfolder named after package

pmalic opened the pull request:
Implemention for issue #3706.

master couldn't compile (using OpenJDK 11.0.2), had to change JdkXmlParsers.scala first (commit pmalic@96d6929).

Also changed string interpolators for `requestBodyFileName`/`responseBodyFileName` to `s`, it was `f` although there were no printf-style format strings.

pmalic commented:
Simulation files now have a subfolder referenced in RawFileBody(), for example:
```scala
.body(RawFileBody("pkg3/cls3_0012_request.txt"))
.check(bodyBytes.is(RawFileBody("pkg3/cls3_0012_response.txt")))
```

The separator in the string argument is hardcoded to `/`, `File.separator` is not used here because in this case a simulation recorded on Windows wouldn't work on Unixes, and `/` works on Windows. There are operating systems where `/` wouldn't work (e.g. OpenVMS), not sure would this be an issue for Gatling. 

I've considered other options, like making the simulation "runner" (don't know the correct term, looking into Gatling code for the first time) read from the subfolder by default, but this would break compatibility, people would have to move their old resources to subfolders. Another option I've considered is generating some code in the simulation file that would make the simulation "runner" look for resources in a subfolder, but to prevent again hardcoding the subfolder path, it would have to use reflection to figure out the package. This is beyond my current very limited Scala skills :)

slandelle commented:
> master couldn't compile (using OpenJDK 11.0.2), had to change JdkXmlParsers.scala first (commit pmalic/gatling@96d6929).

Gatling has to be compiled with JDK8. Compiling with JDK11 would produce bytecode that can't run on JDK8 (typically, because of `ByteBuffer#flip` that would cause LinkageError).
So this change is a no go.

> The separator in the string argument is hardcoded to / ..., not sure would this be an issue for Gatling.

It's OK because we mostly use classloader resolution.

slandelle commented on gatling-core/src/main/scala/io/gatling/core/check/extractor/xpath/JdkXmlParsers.scala:
```diff
@@ -74,7 +74,7 @@ class JdkXmlParsers(configuration: GatlingConfiguration) {
         override def getPrefix(uri: String): String = throw new UnsupportedOperationException
 
-        override def getPrefixes(uri: String): java.util.Iterator[_] = throw new UnsupportedOperationException
+        override def getPrefixes(uri: String): java.util.Iterator[String] = throw new UnsupportedOperationException
```
Please revert

pmalic replied:
Reverted

slandelle commented:
Moreover, could you please sign our contributor license agreement? https://docs.google.com/forms/d/1hbpxVqJ5hIYYJuOjOyOzub9fJFGFGvvfnB5w5LCWdUw

pmalic commented:
Signed contributor license agreement.

slandelle commented:
Thanks!

Actually, I feel like only naming after the package is not enough.
We probably should turn the classname into a folder too (lower case), eg:

`foo.bar.BazSimulation` => `foo/bar/bazsimulation`

WDYT?

pmalic commented:
You already have resource file names prefixed with the class name, so you won't get a "conflict" with multiple classes in one package. The benefit would be to not have possibly thousands of files in one folder if you have multiple classes in a package and each class with a lot of resource files. Yeah, I think it makes sense to introduce another subfolder named after the class.

slandelle commented:
> Yeah, I think it makes sense to introduce another subfolder named after the class. Then the class name prefix in the resource file names should be dropped?

Yes, please.

pmalic commented:
Ok, it's now:
```scala
.body(RawFileBody("com/example/MySim/0005_request.txt"))
.check(bodyBytes.is(RawFileBody("com/example/MySim/0005_response.txt")))
```

pmalic commented:
I've also now implemented changes that set correct file extension to resource files so it's not always `.txt`. This makes opening them a lot easier. I extract mime type from `Content-Type` header. I have a map for a dozen most common mime types and for the rest `jodd.net.MimeTypes.findExtensionsByMimeTypes()` is used (it's based on Apache's `mime.types` file). I have a local map because this `MimeTypes` is not optimized for looking up extension from mime type but vice-versa (looked at the source code, they have a `HashMap` from extension to mime type). The changes touch the code I've already changed in this pull request and they are kind of related to this. Should I commit the changes to my branch associated with this pull request or should I make another one?

slandelle commented:
@pmalic No big deal, but those last changes regarding file extensions should have gone into a separate PR.

We're in the process of releasing, so we'll only be able to merge after that, all the more as I'm a bit concerned with the file path change on Windows that will require proper testing.

Thanks for your patience.

pmalic commented:
I didn't commit the changes related to file extensions to the branch associated with this PR, it's in a separate branch, I just pasted the link to the commit.

slandelle commented:
Ah, ok ,thanks!

pmalic commented:
Created PR #3711 for file extensions.

P.S. You mentioned testing on Windows, I'm mostly running simulations on Windows (I build gatling on macOS, but the app I'm load testing is on Windows).

slandelle commented:
@Neoware can you have a look please?

Neoware commented:
Hey @pmalic, this looks good except for the semicolon at the end of L56 :)

pmalic commented:
Thanks @Neoware for the review, will fix. I'll also squash the current commit with the fix, OK?

Neoware commented:
yes please :)

pmalic commented:
Done.

Neoware commented:
Your commit is merged :) Thank you for you contribution! Let's jump to the next one now :)
